torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,215 @@
+#include <ATen/core/jit_type.h>
+#include <jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/integer_value_refinement.h>
+#include <torch/csrc/jit/passes/value_refinement_utils.h>
+#include <torch/csrc/utils/memory.h>
+
+namespace torch {
+namespace jit {
+
+using IntegerRefinement = std::unordered_map<Value*, int64_t>;
+
+// see [value refinement algorithm] for full explanation.
+// When a comparison like `cond = x == 4` or `cond = x != 4` is made,
+// `cond` value carries information (refinements) about the value of `x`.
+// in an example like:
+// if x == 1:
+//    ...
+// we can substitute all uses of x dominated by the true block
+// with 1.
+
+struct IntegerValueRefiner {
+  IntegerValueRefiner(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
+    if (!checkForPossibleRefinements(graph_->block())) {
+      return false;
+    }
+    IntegerRefinement refinements;
+    RefineIntegerValues(graph_->block(), refinements);
+    return changed_;
+  }
+
+  bool checkForPossibleRefinements(Block* b) {
```

Comment:
Nit: name can be made more specific, like "blockHasIntComparisons". Up to you of course, but "possible refinements" might sound a bit too vague.

Reply:
`blockHasIntComparisons` does describe the current pass, but it also doesn't leave room for improvement, e.g. I could add checks that either value in the int comparison has subsequent uses

Reply:
When an improvement is made, we can rename it to reflect the new functionality :)